test/gst-msdk/vpp/csc.py

```diff
@@ -0,0 +1,54 @@
+###
+### Copyright (C) 2018-2019 Intel Corporation
+###
+### SPDX-License-Identifier: BSD-3-Clause
+###
+
+from ....lib import *
+from ..util import *
+
+spec = load_test_spec("vpp", "csc")
+
+@slash.requires(have_gst)
+@slash.requires(*have_gst_element("msdk"))
+@slash.requires(*have_gst_element("msdkvpp"))
+@slash.requires(*have_gst_element("checksumsink2"))
+@slash.requires(using_compatible_driver)
+@slash.parametrize(*gen_vpp_csc_parameters(spec))
+@platform_tags(VPP_PLATFORMS)
+def test_default(case, csc):
+  params = spec[case].copy()
+  params.update(
+    csc = csc, mcscu = mapformatu(csc),
+    mformat = mapformat(params["format"]))
+  params["ofile"] = get_media()._test_artifact(
+    "{}_{format}_csc_{csc}_{width}x{height}"
+    ".yuv".format(case, **params))
+
+  call(
+    "gst-launch-1.0 -vf filesrc location={source} num-buffers={frames}"
+    " ! rawvideoparse format={mformat} width={width} height={height}"
+    " ! msdkvpp hardware=true ! videoconvert ! video/x-raw,format={mcscu}"
+    " ! checksumsink2 file-checksum=false frame-checksum=false"
+    " plane-checksum=false dump-output=true dump-location={ofile}"
+    "".format(**params))
+
+  if same_pixel_subsample(params["format"], params["csc"]):
+    ssim = calculate_ssim(
+    params["source"], params["ofile"],
+    params["width"], params["height"], params["frames"],
+    params["format"], params["csc"])
+
+    get_media()._set_test_details(ssim = ssim)
+    assert sum(ssim) == 6.0, "Y, U and V values should not be changed
+      by CSC if the input and output have same pixel subsample"
+  else:
+    ssim = calculate_ssim(
+    params["reference"], params["ofile"],
+    params["width"], params["height"], params["frames"],
+    params["csc"])
+
+    get_media()._set_test_details(ssim = ssim)
+    assert 1.0 >= ssim[0] >= 0.99
+    assert 1.0 >= ssim[1] >= 0.99
+    assert 1.0 >= ssim[2] >= 0.99
```

Comment:
I thought we decided this should be user configurable (i.e. the metric).  Let the user set the reference file.  The test should not need to check if subsampling is the same or not.  It should be up to the user to set the reference correctly (i.e. one with same subsample as the csc output).

Reply:
Thus, this should work:
```python
check_metric(
  # if user specified metric, then use it.  Otherwise, use ssim metric with perfect score
  metric = params.get("metric", dict(type = "ssim", miny = 1.0, minu = 1.0, minv = 1.0)),
  # If user specified reference, use it.  Otherwise, assume source is the reference.
  reference = params.get("reference", params["source"]),
  decoded = params["ofile"],
  # if user specified reference, then assume it's format is the same as csc output format.
  # Otherwise, the format is the source format
  format = params["format"] if params.get("reference", None) is None else params["csc"],
  format2 = params["csc"],
  width = params["width"], height = params["height"], frames = params["frames"],
)

Reply:
During modifying the code, I thought a potential problem which the user don't set reference and input/output are different subsampling, then the result will fail, so I add the condition again.
How about
```
if same_pixel_subsample(params["format"], params["csc"]) and (not params.get("reference")):
  params["source"], params["ofile"]
else:
  check_metric(
    metric = params.get("metric", dict(type = "ssim", miny = 1.0, minu = 1.0, minv = 1.0)),
    decoded = params["ofile"],
    format = params["csc"],
    width = params["width"], height = params["height"], frames = params["frames"],
)
```

Reply:
> During modifying the code, I thought a potential problem which the user don't set reference and input/output are different subsampling, then the result will fail, so I add the condition again.

Then that is user error and clearly should go back and define the test case properly.

> How about
> ```
> if same_pixel_subsample(params["format"], params["csc"]) and (not params.get("reference")):
>   check_metric(
>     metric = params.get("metric", dict(type = "ssim", miny = 1.0, minu = 1.0, minv = 1.0)),
>     decoded = params["ofile"],
>     reference = params["source"],
>     format = params["format"],
>     format2 = params["csc"],
>     width = params["width"], height = params["height"], frames = params["frames"],
>   )
> else:
>   check_metric(
>     metric = params.get("metric", dict(type = "ssim", miny = 1.0, minu = 1.0, minv = 1.0)),
>     decoded = params["ofile"],
>     reference = params["reference"],
>     format = params["csc"],
>     width = params["width"], height = params["height"], frames = params["frames"],
>   )
> ```
> should be like this

I think this is unnecessary complexity.  We don't need same_pixel_subsample since it is up to the user to do define the case properly.  Please go with my previous solution.

Reply:
It's not like failing cases will go unnoticed.